gcc/rust/checks/errors/borrowck/ffi-polonius/src/lib.rs

```diff
@@ -172,9 +198,78 @@ pub unsafe extern "C" fn polonius_run(
         }
     }
 
+    let loan_errors = FFIVector__new_vec_pair(output.errors.len());
+    let mut buffer_string = "Loan errors :".to_string();
+    for (keys, values) in output.errors.iter() {
+        let point: usize = (*keys).into();
+        buffer_string.push_str("\nPoint : ");
+        buffer_string.push_str(&point.to_string());
+
+        let loans_vec = FFIVector__new(values.len());
+        for loan in values.iter() {
+            buffer_string.push_str("\nLoan : ");
+            let loan: usize = (*loan).into();
+            buffer_string.push_str(&loan.to_string());
+
+            FFIVector__push(loans_vec, loan);
+        }
+        let pair = gccrs_ffi::Pair {
+            first: point,
+            second: loans_vec,
+        };
+        FFIVector__push_vec_pair(loan_errors, pair);
+    }
+
+    let move_errors = FFIVector__new_vec_pair(output.move_errors.len());
+    buffer_string.push_str("\n\nMove errors :");
+    for (keys, values) in output.move_errors.iter() {
+        let point: usize = (*keys).into();
+        buffer_string.push_str("\nPoint : ");
+        buffer_string.push_str(&point.to_string());
+
+        let paths_vec = FFIVector__new(values.len());
+        for path in values.iter() {
+            buffer_string.push_str("\nPath : ");
+            let path: usize = (*path).into();
+            buffer_string.push_str(&path.to_string());
+
+            FFIVector__push(paths_vec, path);
+        }
+        let pair = gccrs_ffi::Pair {
+            first: point,
+            second: paths_vec,
+        };
+        FFIVector__push_vec_pair(move_errors, pair);
+    }
+
+    let subset_errors = FFIVector__new_vec_triple(output.subset_errors.len());
+    buffer_string.push_str("\n\nSubset errors :");
+    for (key, value) in output.subset_errors.iter() {
+        let point: usize = (*key).into();
+        buffer_string.push_str("\nPoint : ");
+        buffer_string.push_str(&point.to_string());
+        for origin_pair in value.iter() {
+            let origin_1: usize = origin_pair.0.into();
+            let origin_2: usize = origin_pair.1.into();
+            buffer_string.push_str("\nOrigin1 : ");
+            buffer_string.push_str(&origin_1.to_string());
+            buffer_string.push_str("\nOrigin2 : ");
+            buffer_string.push_str(&origin_2.to_string());
+            let triple = gccrs_ffi::Triple {
+                first: point,
+                second: origin_1,
+                third: origin_2,
+            };
+            FFIVector__push_vec_triple(subset_errors, triple);
+        }
+    }
+
+    // write a dump for polonius output to verify data is correctly moved to C++ side
+    use std::fs;
+    fs::write("rust_side_polonius_output.txt", buffer_string).expect("Error writing to file!");
```

Comment:
hm, I would remove the file output here and all of the operations on `buffer_string`. I think it is a good way of debugging this (and I'm glad you thought about it, I think it was the correct move) but it makes the final code very hard to read. I think we should be able to find memory issues by running the compiler with ASAN, so I'm not sure on how useful this will be.

if you think it is worth keeping, then I would put it in a completely different function. both functions could iterate on the errors, and one could build the `buffer_string` while the other could build the resulting `loan_errors`, `move_errors` and `subset_errors` vectors

Reply:
I think I won't need really need this, will get rid of the debugging code